automates/program_analysis/CAST2GrFN/visitors/cast_to_agraph_visitor.py

```diff
@@ -0,0 +1,449 @@
+        """Visits Attribute nodes, the node's UID is returned
+        so it can be used to connect nodes in the digraph"""
+        value = self.visit(node.value)
+        attr = self.visit(node.attr)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label="Attribute")
+        self.G.add_edge(node_uid,value)
+        self.G.add_edge(node_uid,attr)
+        
+        return node_uid
+
+    @visit.register
+    def _(self, node: BinaryOp):
+        """Visits BinaryOp nodes, the node's UID is returned
+        so it can be used to connect nodes in the digraph"""
+        left = self.visit(node.left)
+        right = self.visit(node.right)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label=node.op)
+        self.G.add_edge(node_uid, left)
+        self.G.add_edge(node_uid, right)
+
+        return node_uid
+
+    @visit.register
+    def _(self, node: Call):
+        """Visits Call (function call) nodes. We check to see
+        if we have arguments to the node and act accordingly. 
+        Appending all the arguments of the function to this node,
+        if we have any. The node's UID is returned."""
+        func = self.visit(node.func)
+        args = []
+        if(node.arguments != []):
+            args = self.visit_list(node.arguments)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label="FunctionCall")
+        self.G.add_edge(node_uid, func)
+        for n in args:
+            self.G.add_edge(node_uid,n)
+
+        return node_uid
+
+    @visit.register
+    def _(self, node: ClassDef):
+        """Visits ClassDef nodes. We visit all fields and functions
+        of the class definition, and connect them to this node.
+        This node's UID is returned."""
+        # TODO: Where should bases field be used?
+        funcs = []
+        fields = [] 
+        if(node.funcs != []):
+            funcs = self.visit_list(node.funcs)
+        if(node.fields != []):
+            fields = self.visit_list(node.fields)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label="Class: " + node.name)
+        attr_uid = uuid.uuid4()
+        self.G.add_node(attr_uid,label="Attributes")
+        self.G.add_edge(node_uid,attr_uid)
+        for n in funcs + fields:
+            self.G.add_edge(attr_uid,n)
+
+        return node_uid
+
+    @visit.register
+    def _(self, node: Dict):
+        """Visits Dictionary nodes. We visit all the keys and values of
+        this dictionary and then they're added to this node. Right now 
+        they're just added by adding keys and then values.
+        This node's UID is returned."""
+        keys = []
+        values = []
+        if(node.keys != []):
+            keys = self.visit_list(node.keys)
+        if(node.values != []):
+            values = self.visit_list(node.values)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label="Dict")
+
+        # TODO: Make this prettier, somehow?
```

Comment:
this loop looks fine to me :)

Reply:
@titomeister the function you are looking for is `add_edges_from` (documented [here](https://networkx.org/documentation/stable//reference/classes/generated/networkx.Graph.add_edges_from.html)). You can combine that with a [generator expression](https://docs.python.org/3/howto/functional.html#generator-expressions-and-list-comprehensions) to form the following Pythonic statement:
```python
self.G.add_edges_from((node_uid, n) for n in keys + values)
```

Reply:
Thanks Daniel :--) 
Thank you Paul! I'm going to add this later because I want to experiment with it a little bit before committing it to the code. I made myself a note about it so I won't forget. I'll go ahead and remove the TODO comment from here for now though.